metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py

```diff
@@ -0,0 +1,687 @@
+                "Denied permission for read text from procedure '%s'",
+                procedure.full_name,
+            )
+            return None, None
+        code_list = []
+        code_slice_index = 0
+        code_slice_text = "create procedure"
+        try:
+            for index, row in enumerate(code_data):
+                code_list.append(row["Text"])
+                if code_slice_text in re.sub(" +", " ", row["Text"].lower()).strip():
+                    code_slice_index = index
+            definition = "\n".join(code_list[:code_slice_index])
+            code = "\n".join(code_list[code_slice_index:])
+        except ResourceClosedError:
+            logger.warning(
+                "Connection was closed from procedure '%s'",
+                procedure.full_name,
+            )
+            return None, None
+        return definition, code
+
+    @staticmethod
+    def _get_procedure_properties(
+        conn: Connection, procedure: StoredProcedure
+    ) -> Dict[str, Any]:
+        properties_data = conn.execute(
+            f"""
+            SELECT
+                create_date as date_created,
+                modify_date as date_modified
+            FROM sys.procedures
+            WHERE object_id = object_id('{procedure.full_name}')
+            """
+        )
+        properties = {}
+        for row in properties_data:
+            properties = dict(
+                date_created=row["date_created"], date_modified=row["date_modified"]
+            )
+        return properties
+
+    @staticmethod
+    def _get_stored_procedures(
+        conn: Connection, db_name: str, schema: str
+    ) -> List[Dict[str, str]]:
+        stored_procedures_data = conn.execute(
+            f"""
+            SELECT
+                pr.name as procedure_name,
+                s.name as schema_name
+            FROM
+                [{db_name}].[sys].[procedures] pr
+            INNER JOIN
+                [{db_name}].[sys].[schemas] s ON pr.schema_id = s.schema_id
+            where s.name = '{schema}'
+            """
+        )
+        procedures_list = []
+        for row in stored_procedures_data:
+            procedures_list.append(
+                dict(db=db_name, schema=row["schema_name"], name=row["procedure_name"])
+            )
+        return procedures_list
+
+    def construct_job_workunits(
+        self,
+        data_job: MSSQLDataJob,
+    ) -> Iterable[MetadataWorkUnit]:
+        mcp = MetadataChangeProposalWrapper(
+            entityType=data_job.type,
+            entityUrn=data_job.urn,
+            changeType=ChangeTypeClass.UPSERT,
+            **data_job.as_datajob_info_aspect_data,
+        )
+        wu = MetadataWorkUnit(
+            id=f"{data_job.source}.{data_job.entity.full_name}.{mcp.aspectName}",
+            mcp=mcp,
+        )
+        self.report.report_workunit(wu)
+        yield wu
+
+        mcp = MetadataChangeProposalWrapper(
+            entityType=data_job.type,
```

Comment:
entityType, aspectName, and changeType are not required

can we change as_datajob_input_output_aspect_data to return the aspect directly instead of a dict?

Reply:
done